crates/matrix-sdk-sqlite/src/event_cache_store.rs

```diff
@@ -0,0 +1,385 @@
+    fs::create_dir_all(path).await.map_err(OpenStoreError::CreateDir)?;
+    let cfg = deadpool_sqlite::Config::new(path.join("matrix-sdk-media.sqlite3"));
+    Ok(cfg.create_pool(Runtime::Tokio1)?)
+}
+
+/// Initialize the database.
+async fn init(conn: &SqliteConn) -> Result<()> {
+    // First turn on WAL mode, this can't be done in the transaction, it fails with
+    // the error message: "cannot change into wal mode from within a transaction".
+    conn.execute_batch("PRAGMA journal_mode = wal;").await?;
+    conn.with_transaction(|txn| {
+        txn.execute_batch(include_str!("../migrations/event_cache_store/001_init.sql"))
+    })
+    .await?;
+
+    conn.set_kv("version", vec![1]).await?;
+
+    Ok(())
+}
+
+#[async_trait]
+trait SqliteObjectEventCacheStoreExt: SqliteObjectExt {
+    async fn set_media(&self, uri: Key, format: Key, data: Vec<u8>) -> Result<()> {
+        self.execute(
+            "INSERT OR REPLACE INTO media (uri, format, data, last_access) VALUES (?, ?, ?, CAST(strftime('%s') as INT))",
+            (uri, format, data),
+        )
+        .await?;
+        Ok(())
+    }
+
+    async fn get_media(&self, uri: Key, format: Key) -> Result<Option<Vec<u8>>> {
+        Ok(self
+            .with_transaction::<_, rusqlite::Error, _>(move |txn| {
+                let Some(media) = txn
+                    .query_row::<Vec<u8>, _, _>(
+                        "SELECT data FROM media WHERE uri = ? AND format = ?",
+                        (&uri, &format),
+                        |row| row.get(0),
+                    )
+                    .optional()?
+                else {
+                    return rusqlite::Result::Ok(None);
+                };
+
+                // Update the last access.
+                txn.execute(
+                    "UPDATE media SET last_access = CAST(strftime('%s') as INT) WHERE uri = ? AND format = ?",
+                    (uri, format),
+                )?;
+
+                rusqlite::Result::Ok(Some(media))
+            })
+            .await?)
+    }
+
+    async fn remove_media(&self, uri: Key, format: Key) -> Result<()> {
+        self.execute("DELETE FROM media WHERE uri = ? AND format = ?", (uri, format)).await?;
+        Ok(())
+    }
+
+    async fn remove_uri_medias(&self, uri: Key) -> Result<()> {
+        self.execute("DELETE FROM media WHERE uri = ?", (uri,)).await?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl SqliteObjectEventCacheStoreExt for deadpool_sqlite::Object {}
+
+#[async_trait]
+impl EventCacheStore for SqliteEventCacheStore {
+    type Error = Error;
+
+    async fn add_media_content(&self, request: &MediaRequest, content: Vec<u8>) -> Result<()> {
+        let uri = self.encode_key(keys::MEDIA, request.source.unique_key());
+        let format = self.encode_key(keys::MEDIA, request.format.unique_key());
+        let data = self.encode_value(content)?;
+        self.acquire().await?.set_media(uri, format, data).await
```

Comment:
Out of curiosity, what's the reasoning behind having multiple layers to perform SQL queries? (i.e. having `set_media` somewhere else, and not inlined here)

In my recent works on the send queue, I've put all the code at the same layer, and it feels simpler to understand, and with fewer indirections.

Reply:
I am not sure anymore. I mostly copy-pasted here what was in the `StateStore` implementation so we can definitely reduce the number of layers.

Reply:
Good, we can do this in a follow-up PR!